sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memory_alloc.asciidoc

```diff
@@ -0,0 +1,755 @@
+
+  initial_threshold_key(size_t initialThreshold);
+};
+
+struct maximum_size_key {
+  using value_t = property_value<maximum_size_key>;
+
+  initial_threshold_key(size_t maxSize);
+};
+
+struct read_only_key {
+  using value_t = property_value<read_only_key>;
+
+  read_only_key(bool readOnly);
+};
+
+struct zero_init_key {
+  using value_t = property_value<zero_init_key>;
+
+  read_only_key(bool zeroInit);
+};
+
+struct use_existing_memory_key {
+  using value_t = property_value<use_existing_memory_key>;
+
+  use_existing_memory(void *ptr, size_t size);
+};
+
+inline constexpr initial_threshold_key::value_t initial_threshold;
+inline constexpr maximum_size_key::value_t maximum_size;
+inline constexpr read_only_key::value_t read_only;
+inline constexpr zero_init_key::value_t zero_init;
+inline constexpr use_existing_memory_key::value_t use_existing_memory;
+
+}  // ext::oneapi::experimental
+----
+
+|===
+|Property|Description
+
+|`initial_threshold`
+|The `initial_threshold` property specifies the initial deallocation threshold
+ value for the memory pool. If this property is not used the default value is
+ zero, and this can be increased after the memory pool is created by calling
+ `memory_pool::set_new_threshold`.
+
+|`maximum_size`
+|The `maximum_size` property specifies the maximum size of the memory pool,
+ after which any allocation will result in an exception. If the value specified
+ is larger than the implementation can support an exception with the
+ `errc::memory_allocation` error code is thrown. If this property is not used
+ the default value is implementation-defined.
+
+|`read_only`
+|The `read_only` property is a performance hint which asserts that all memory
+ allocations from the memory pool will only ever be read from, this can be used
+ by the SYCL runtime to optimize for performance.
+
+|`zero_init`
+|The `zero_init` property adds the requirement that all memory allocated to the
+ memory as it is allocated from the source to the memory pool will be
+ initialised to zero. Note there is no guarantee that the memory allocation be
+ re-initialized to zero when it is re-allocated from the pool, so users must
+ re-initialize memory to zero if they wish for later allocations to have this
+ behavior.
+
+|`use_existing_memory`
+|The `use_existing_memory` property adds the requirement that the memory pool
+ will use an existing USM memory allocation provided by the user instead of
+ allocating from the OS. This property takes a pointer to a valid USM memory
+ allocation of the same allocation kind as the memory pool is initialized with
+ and the size of that memory allocation. Using this property will implicitly set
+ the `maximum_size` and `initial_threshold` property to be that of the size
+ provided, and as such using the `maximum_size` or `initial_threshold`
+ properties in conjunction with this property will cause the `memory_pool`
+ constructor to throw an exception with the `errc::invalid` error code.
```

Comment:
I like the `use_existing_memory` concept, but I wonder if it should be exposed as a new constructor instead.  There is no need for the pool to know the USM memory "kind" if the user provides their own USM memory, so the constructor could be simplified to:

```
template <typename Properties = empty_properties_t>
memory_pool(context ctx, void* ptr, size_t size, Properties props = {});
```

However, this makes it hard to implement `get_alloc_kind` in this case.  Is that query really useful, though?  I think I'd add the constructor I list above and also remove `get_alloc_kind`.

Reply:
I agree it would be better to expose this as an alternative constructor, I've switched from the property to a constructor like what you proposed in https://github.com/intel/llvm/commit/48493ace0babceba6b65fe7399b43284c10681db.

Though I don't think this necessarily makes it difficult to implement `get_alloc_kind`, I think the constructor which takes an existing USM memory allocation could infer the allocation kind by calling `get_pointer_type`.

Reply:
> Though I don't think this necessarily makes it difficult to implement `get_alloc_kind`, I think the constructor which takes an existing USM memory allocation could infer the allocation kind by calling `get_pointer_type`

Good point.